core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -984,6 +984,11 @@ class DAGScheduler(
                   job.numFinished += 1
                   // If the whole job has finished, remove it
                   if (job.numFinished == job.numPartitions) {
+                    // Clear failure count for this stage, now that it's succeeded. 
+                    // This ensures that even if subsequent stages fail, triggering 
+                    // a recompute of this stage, we abort because of those failures. 
+                    stage.clearFailures()
```

Comment:
this is only getting applied to the `ResultStages`, but it should get applied to `ShuffleMapStage` as well.  In fact I'd suggest moving it to the success branch in `markStageAsFinished`.